src/dists.ad.js

```diff
@@ -1347,14 +1351,90 @@ var Marginal = makeDistributionType({
     return this.supp;
   },
   print: function() {
-    return _.map(this.params.dist, function(obj, val) { return [val, obj.prob]; })
-        .sort(function(a, b) { return b[1] - a[1]; })
-        .map(function(pair) { return '    ' + pair[0] + ' : ' + pair[1]; })
-        .join('\n');
+    return printDist(this.params.dist);
   }
 });
 
 
+// A "list of samples" backed marginal distribution that only
+// aggregates the samples into a distribution when necessary.
+var SampleBasedMarginal = makeDistributionType({
+  name: 'Marginal',
+  nodoc: true,
+  nohelper: true,
+  params: [{name: 'samples'}],
+  mixins: [finiteSupport],
+  constructor: function() {
+    if (!_.isArray(this.params.samples) ||
+        this.params.samples.length === 0) {
+      throw new Error('Expected samples to be a non-empty array.');
+    }
+
+    // Provide access to the samples using the interface previously
+    // provided by the `justSample` option.
+    // samples is an array of objects like: {value: ..., score: ...}
+    this.samples = this.params.samples;
+
+    this.getDist = function() {
+      if (this._cacheddist) {
+        return this._cacheddist;
+      } else {
+        var dist = {};
+        this.params.samples.forEach(function(obj) {
+          var val = obj.value;
+          var k = util.serialize(val);
+          if (dist[k] === undefined) {
+            dist[k] = {val: val, prob: 0};
+          }
+          dist[k].prob += 1;
+        });
+        // Normalize.
+        var n = this.params.samples.length;
+        _.each(dist, function(obj) { obj.prob /= n; });
+        this._cacheddist = dist;
+        return dist;
+      }
+    };
+
+    this.getSupport = function() {
+      if (this.params.samples.length === 1) {
+        // Optimization: Avoid unnecessary serialization in the
+        // onlyMAP case
+        return [this.params.samples[0].value];
+      } else if (this._cachedsupport) {
+        return this._cachedsupport;
+      } else {
+        var support = _.map(this.getDist(), _.property('val'));
+        this._cachedsupport = support;
+        return support;
+      }
+    };
+  },
+  sample: function() {
+    var n = this.params.samples.length;
+    return this.params.samples[Math.floor(util.random() * n)].value;
+  },
+  score: function(val) {
+    var k = util.serialize(val);
+    var obj = this.getDist()[k];
+    return (obj === undefined) ? -Infinity : Math.log(obj.prob);
+  },
+  support: function() {
+    return this.getSupport();
```

Comment:
What's the reason for having a separate `getSupport` function (as opposed to inlining the code there into the `support` method)?